---
title: "BME4550_Fall2018_Assignment9"
author: "Lindsay McDonald"
date: "November 7, 2018"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(rafalib)
library(UsingR)
library(dplyr)
library(contrast)
library(devtools)
library(GSE5859Subset)
library(genefilter)
library(qvalue)
library(SpikeInSubset)
library(tissuesGeneExpression)
library(RColorBrewer) 
library(gplots)
library(matrixStats)
library(caret)
library(class)
library(dagdata)
library(Biobase)
library(GSE5859)
library(limma)
library(sva)
```

## Batch Effects

### Exercises 1

#### 1. What is the proportion of women that were accepted?
```{r}
data(admissions)
# men
index = which(admissions$Gender==1)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied

# women
index = which(admissions$Gender==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```
The proportion of women accepted was 30.33%

#### 2. Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value? This difference actually led to a lawsuit. Now notice that looking at the data by major, the differences disappear. How can this be? This is referred to as Simpson's Paradox
```{r}
atab<-as.data.frame(matrix(0,2,2))
rownames(atab)<-c("Men","Women")
colnames(atab)<-c("Accepted","Rejected")
index = admissions$Gender==1
men=admissions[index,]
women=admissions[!index,]
atab[1,1]= sum(men$Number * men$Percent/100)
atab[1,2]= sum(men$Number*(1-men$Percent/100))
windex = which(admissions$Gender==0)
atab[2,1]= sum(women$Number * women$Percent/100)
atab[2,2]= sum(women$Number * (1-women$Percent/100))
atab
chisq.test(atab)$p.value

# looking by major 
index = admissions$Gender==1
men = admissions[index,]
women = admissions[!index,]
print( data.frame( major=admissions[1:6,1],men=men[,3], women=women[,3]) )
```
The p-value is 9.14 e -22 which is considered significant.

#### 3.  We can quantify how "hard" a major is by using the percent of students that were accepted. Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major?
```{r}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
H
major[which.min(H)]
```
The hardest is major is F

#### 4. What proportion is accepted for this major?
```{r}
min(H)
```
6.477% were accepted into Major F

#### 5. . For men, what is the correlation between the number of applications across majors and H?
```{r}
cor(H,men$Number)
```

#### 6. For women, what is the correlation between the number of applications across majors and H?
```{r}
cor(H,women$Number)
```

#### 7. Given the answers to the above, which best explains the differences in admission percentages when we combine majors?
There is confounding between gender and preference for "hard" majors: females are more likely to apply to harder majors

### Exercises 2

#### 1. Familiarize yourself with the sampleInfo table. Note that ethnic group and year is almost perfectly confounded:
```{r}
data(GSE5859)
geneExpression = exprs(e)
sampleInfo = pData(e)

year = format(sampleInfo$date,"%y")
length( unique(year) )
table(year, sampleInfo$ethnicity)
```

#### 2. For how many of these years do we have more than one ethnicity represented.
```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```

#### 3. Repeat the above exercise, but now, instead of year, consider the month as well. For what proportion of these month.year values do we have more than one ethnicity represented?
```{r}
month.year = format(sampleInfo$date,"%m%y")

mytab=table(month.year,sampleInfo$ethnicity)
print(mytab)
myx=rowSums(mytab!=0)
mean(myx>=2)
```
only 4.76% of the months had more than one ethnicity represented.

#### 4. . Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05 ?
```{r}
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### 5. 5. What is the estimate of pi0 provided by qvalue:
```{r}
qval$pi0
```

#### 6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values less than 0.05?
```{r}
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### 7. Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population tothe CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05 ?
```{r}
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### 8. Over 80% of genes are called differentially expressed between ethnic groups. How many genes have q-values < 0.05
```{r}
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

#### 9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3) How many genes have q-values < 0.05 ?
```{r}
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

### Exercises 3

#### 1. Using the functions rowttests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1?
```{r}
data(GSE5859Subset)

sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m"))
#table( sampleInfo$group, month)

sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.1)
```

#### 2. Note that sampleInfo$group here presents males and females. Thus, we expect differences to be in on chrY and, for genes that escape inactivation, chrX For the list calculated above, what proportion of this list is on chrX or chrY?
```{r}
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

#### 3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrY")
sum(index=="TRUE")
```

#### 4. Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1, perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values <0.05 ? 
```{r}
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = which(qvals<0.1 & !geneAnnotation$CHR%in%c("chrX","chrY"))
month = factor( format(sampleInfo$date,"%m"))
pval = rowttests(geneExpression[index,], month)$p.value
mean(pval<0.05)
```

#### 5. The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect.
```{r}
(X = model.matrix(~sex+month))
```

#### 6. Now use the X defined above, to fit a regression model using lm for each gene. You can obtain p-values for estimated parameters using  How many of the q-values for the group comparison are now <0.1?
```{r}
X = model.matrix(~sex+month)
pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[2,4]
  # the 2 picks out the second (sex) coef, the 4 picks out the p value
} ) )
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)

```

#### 7. With this new list, what proportion of these are chrX and chrY? Notice the big improvement.
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

#### 8. How many on Y or X?
```{r}
sum(index)
```

#### 9. Now from the linear model above, extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1? 
```{r}
X = model.matrix(~sex+month)
pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[3,4] # third coef, for month
} ) )
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

### Exercises #4 

#### 1. Suppose you want to make an MA plot of the first two samples y = geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot?
B) y(1 1; 1 -1)

#### 2. Say Y is MxN, in the SVD, Y = UDV^T which of the following is not correct?
C) D are the coordinates of the projection U^T*Y


#### 3. Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation. The only difference in these plots is the order in which the samples are plotted. Which of the following would you say is true?
```{r}
data(GSE5859Subset)
y = geneExpression - rowMeans(geneExpression)

mypar(1,2)
n=ncol(y)
cors=cor(y)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
o <- order(sampleInfo$date)
yo<-y[,o]
cors=cor(yo)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```
D) The fact that in the plot ordered by month we see two groups mainly driven by month, and within these we see subgroups driven by date, seems to suggest date more
than month per se are the hidden factors.

#### 4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates.
```{r}
s <- svd(y)
What <- s$v[,1:2] # these are the first two PCs.
rownames(What)<-colnames(y)
round(What,2)
fit = s$u[,1:2]%*% (s$d[1:2]*t(What))
var(as.vector(fit))/var(as.vector(y))
```
The command pcs = svd(y)$v[,1:2]


#### 5. Plot each of the estimated factors ordered by date. Use color to denote month. The first factor is clearly related to date. Which of the following appear to be most different according to this factor?
```{r}
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
mypar(2,1)
for(i in 1:2){
  plot(pcs[o,i],col=cols,xaxt="n",xlab="")
  label = gsub("2005-","",sampleInfo$date[o])
  axis(1,1:ncol(y),label,las=2)
}
```
23 June and 27 June

#### 6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?
```{r}
s = svd(y)
varexplained = s$d^2/ sum(s$d^2)
plot(varexplained)
sum(varexplained>0.10)
```

#### 7. Which PC most correlates (negative or positive correlation) with month?
```{r}
s = svd(y)
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),s$v)
plot(t(cors))
which.max(abs(cors))
```

#### 8. What is this correlation (in absolute value)?
```{r}
max(abs(cors))
```

#### 9. Which PC most correlates (negative or positive correlation) with sex?
```{r}
s = svd(y)
sex = sampleInfo$group
cors = cor( as.numeric(sex),s$v)
plot(t(cors))
which.max(abs(cors))
```

#### 10. What is this correlation (in absolute value)?
```{r}
max(abs(cors))
```

#### 11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?
```{r}
X <- model.matrix(~sex+s$v[,1:2])
fit <- lmFit(y,X)
k <- 2 ##second coef, for month
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

#### 12. What proportion of the genes are on chromosomes X and Y?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

### Exercises 5
```{r}
data(GSE5859Subset)
```

#### 1. Now fit a linear model to each gene that instead of month includes these factors in the model. Use the qvalue function.
```{r}
s <- svd(geneExpression-rowMeans(geneExpression))
cor(sampleInfo$group,s$v[,1])

sex = sampleInfo$group
mod = model.matrix(~sex)
svafit = sva(geneExpression,mod)
head(svafit$sv)

for(i in 1:ncol(svafit$sv)){
  print( cor(s$v[,i],svafit$sv[,i]) )
}

svaX<-model.matrix(~sex+svafit$sv)
fit <- lmFit(geneExpression,svaX)
k <- 2 #second coef, for sex
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

#### 2. What proportion of the genes are on chrX or chrY?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

